Fixes and updates to .gitlab-ci.yml

robertkb opened the pull request:
Various tweaks and compiler flag fixes for Lassen builds, and updates to take care of recent changes to the develop branch.

robertkb commented:
All cmake builds are working currently. I'm still working on setting the correct component library versions for the automatic spack testing.

robertkb commented:
All tests pass. Please review and accept, if it looks good.

robertkb commented:
Roland noticed that I am doing some ugly editing of the Soack compilers.yaml file. If there is a better way to automate the editing of the compiler settings, please let me know.

adammoody commented on .gitlab-ci.yml:
```diff
@@ -109,10 +109,9 @@ spack-scr-catalyst:
 #    - git clone -b robertkb-fix-scr-dependencies https://github.com/robertkb/spack.git
     - git clone https://github.com/spack/spack.git
```
Do things still work if you add a ``--depth 1`` here?  If so, let's add that in

adammoody commented on .gitlab-ci.yml:
```diff
@@ -215,11 +214,11 @@ spack-scr-quartz:
     - echo "Working dir " `pwd`
     - rm -rf spack
-    - git clone -b robertkb-fix-scr-dependencies https://github.com/robertkb/spack.git
+    - git clone https://github.com/spack/spack.git
```
Here too, let's see if we can add ``--depth 1`` here.

adammoody commented on .gitlab-ci.yml:
```diff
@@ -320,18 +346,18 @@ spack-scr-lassen:
     - echo "Working dir " `pwd`
-#    - git clone -b robertkb-fix-scr-dependencies https://github.com/robertkb/spack.git
     - rm -rf spack
     - git clone https://github.com/spack/spack.git
```
Another spot, let's see if ``--depth 1`` works, and anywhere else we do a ``git clone``

adammoody commented:
We can look at this for a follow up PR, but what directory to these builds execute in?

If possible, it'd be nice to do the clone and builds in ``/dev/shm`` on the node.  This reduces load on NFS (if that's where the builds are), and it also significantly speeds up the build.

We may still want to install to NFS, though, if we are running parallel tests that use multiple compute nodes.  If not, we could also install to /dev/shm.

robertkb commented:
I updated the spack checkouts to use --depth 1

I modified the cmake builds to create a build dir in /dev/shmem, compile there, then copy those files back to the .gitlab-runner directory for later test stages. One issue is that when there is a problem while building in /dev/shmem, its hard to go back to the failed build and look for log files. For that case, I left a couple of commented-out lines in to set the compile directory to a place in $HOME, so that the files are more easily found if debugging is needed.

The spack tests still run on NFS. I will think about also moving them to /dev/shmem.

The test program compiles and execution are still being run in the build directory on NFS, but those compiles are pretty fast, so its probably not worth moving them. 

The checkout of the cmake SCR source is managed by gitlab, using some sort of incremental checkout. I don't know whether it is efficient or not, but I suspect it is similar to using --depth 1, since the entire existing repo is examined, then only changes are sent.

This can be committed now, and I will use a separate pull request for the spack /dev/shmem build, or I can roll the spack changes into this request as well. I'll continue to update the same branch on gitlab, but won't push it to github until it is in a mergeable state.

adammoody commented:
Looks great.  Thanks @robertkb !
